### PR TITLE
imgcodecs: jpeg: add IMWRITE_JPEG_SAMPLING_FACTOR parameter

### PR DESCRIPTION
**Pull request by Kumataro**
fix https://github.com/opencv/opencv/issues/22052

This merge request contains sample and test program.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Custom=centos:7
buildworker:Custom=linux-f1
```

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_jpeg.cpp [hunk 1]
```diff
@@ -687,6 +688,27 @@ bool JpegEncoder::write( const Mat& img, const std::vector<int>& params )
...
+                    case IMWRITE_JPEG_SAMPLING_FACTOR_420:
+                    case IMWRITE_JPEG_SAMPLING_FACTOR_422:
+                    case IMWRITE_JPEG_SAMPLING_FACTOR_440:
+                    case IMWRITE_JPEG_SAMPLING_FACTOR_444:
+                    // OK.
+                    break;
+
+                    default:
+                    // Unknown parameter is set.
+                    sampling_factor = 0;
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_jpeg.cpp:708]: It makes sense to emit warning here through `CV_LOG_WARNING(NULL, "Unknown value for IMWRITE_JPEG_SAMPLING_FACTOR: " << sampling_factor);`

File: samples/cpp/imgcodecs_jpeg.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#include <vector>
+
+using namespace std;
+using namespace cv;
+
+int main(int /*argc*/, const char** /* argv */ )
+{
+    Mat framebuffer = Mat( 160 * 2, 160 * 5, CV_8UC3, cv::Scalar::all(255) );
+
+    Mat img = Mat( 160, 160, CV_8UC3, cv::Scalar::all(255) );
```
Comment (alalek) [inline samples/cpp/imgcodecs_jpeg.cpp:18]: > `Mat framebuffer = Mat(...);`

No need to copy Mat. Just initialize in ctor: `Mat framebuffer(...);`

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Public API changes are good.

File: modules/imgcodecs/src/grfmt_jpeg.cpp (hunk 1 above)
Comment (Kumataro) [reply modules/imgcodecs/src/grfmt_jpeg.cpp]: Thank you for your comment, i fixed it.

**Comment by Kumataro** [comment]
Thank you for build and test ! I checked compile log. 
For compiling `test_npr.cpp`, W10/BuildContrib is OK, but W10/BuildAndTest was NG,

I think W10/BuildAndTest will become OK if it run again.

[OK] W10 / BuildContrib
- https://github.com/opencv/opencv/runs/6722822212?check_suite_focus=true#step:10:1503

```
[996/2735] Building CXX object modules\ml\CMakeFiles\opencv_test_ml.dir\test\test_mltests.cpp.obj
[997/2735] Building CXX object modules\photo\CMakeFiles\opencv_test_photo.dir\test\test_denoising.cuda.cpp.obj
[998/2735] Building CXX object modules\photo\CMakeFiles\opencv_test_photo.dir\test\test_npr.cpp.obj
[999/2735] Building CXX object modules\photo\CMakeFiles\opencv_test_photo.dir\test\test_denoising.cpp.obj
[1000/2735] Building CXX object modules\photo\CMakeFiles\opencv_test_photo.dir\test\test_decolor.cpp.obj
```

[NG] W10 / BuildAndTest
- https://github.com/opencv/opencv/runs/6722822065?check_suite_focus=true#step:10:1348

```
[1031/1914] Building CXX object modules\photo\CMakeFiles\opencv_test_photo.dir\test\test_npr.cpp.obj
FAILED: modules/photo/CMakeFiles/opencv_test_photo.dir/test/test_npr.cpp.obj 
```

**Comment by Kumataro** [comment]
Thank you for retrying to test. The result of W10/BuildAndTest was success !

**Review by alalek — APPROVED** [review]
LGTM :+1:

**Pull request merged**